### ISSUE TITLE
[Animl Base] Figure out strategy for managing memory and disk space

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-base#20]
We could either (a) delete images immediately once they're uploaded to s3 or (b) set a storage threshold, and once that's reached delete the oldest files as new ones come in.

**Comment by nathanielrindlaub** [comment]
It appears as though we run into memory issues if the directory being watched for new images (`/home/animl/data/<base name>/cameras/`) contains too many images. On the Diablo computer, we [started maxing out](https://github.com/tnc-ca-geo/animl-base/issues/40) the original PM2 `max_memory_refresh` threshold of 1GB once there were ~200k images (totaling 38GB). Once I removed all of the image files, the memory usage dropped to almost nothing.

**Comment by nathanielrindlaub** [comment]
Instead of deleting images we could also just copy them to another directory that's not being watched and that would solve the memory issue but eventually exhaust disk space.

**Comment by postfalk** [comment]
That is an interesting problem. That raises some questions:

1. What is the purpose of keeping the images? Backup? Caching? Etc.
2. Moving them into a different directory would be a good idea.
3. I assume that the stability of the system is more important than keeping old images. But keeping the images would be nice. What is often done in such cases is to delete files randomly. So if your disk usage reaches let's say 90% you might just randomly delete 1 out of five images before a certain date.

**Comment by nathanielrindlaub** [comment]
Good questions @postfalk. The only purpose for retaining them would be for backup I suppose, and if we are certain they made it to S3, I don't know how important that is. 

I like the idea of the random deletion above some threshold, but if we start to have an incomplete backup and we are confident that any images that are queued for deletion are already in Animl, it's a little hard to picture the scenario in which those backup images would come in handy.

A more plausible scenario is that a base station goes offline so it can't upload images for a long time but is still receiving them. In that case we would want to make sure that there is ample disk space and memory to handle a long internet outage, so I guess maintaining a lot of headroom on both counts would be the best strategy.

**Comment by postfalk** [comment]
Agreed. However, a good design would be that would do something sensible when we hit the boundary. In which case it is really a decision. Do we want to have an increasing blurry picture of the past or do we just throw it out in favor of new incoming data.

**Comment by nathanielrindlaub** [comment]
Ok so given that we're going to keep the threshold pretty low (maybe say 25k images), I'm leaning towards just deleting the oldest images once we reach that threshold. If we were to just randomly remove 1 out of every 5 images that would mean we could retain a slightly longer record of the data (i.e. the time extent of the data would be 20% longer) at the cost of that data being 20% blurrier, right?

I don't have strong feelings, but making a hard cutoff and retaining an accurate backup of the the 25k most recent images that were successfully uploaded seems simplest and is a pretty reasonable strategy. What do you think?

**Comment by postfalk** [comment]
Sure. One useful consideration in the math might be that we usually shot more than one image of the same animal. So the information we retain would be still more precise than if the images would be entirely random. BUT I think deleting the oldest ones is sensible as well.

**Comment by nathanielrindlaub** [comment]
Ok after a bit more thought I think this is the path forward I am going to pursue. I think one important thing to note is there are two separate but related problems here: the first being that [chokidar consumes a lot of memory](https://github.com/paulmillr/chokidar/issues/849) as the number of watched files grows, and the second is how to manage available disk space during normal operation (in which images are getting uploaded but we may want to retain a backup of uploaded images) and during internet outages (in which images will pile up on the drive and eventually exhaust the disk space).

I think the following solution would address both:

1. similarly to how we manage images in S3 once they reach the cloud, the images files on the base station would live in one of three directories:
  - an `/ingestion` directory to which new images get written and that's being watched for new files
  - a `/queue` directory which files get moved to as soon as they are detected in `/ingestion` (this would solve the memory issue buy keeping the number of watched files very low)
  - a `/backup` directory to which images that were successfully uploaded get moved
2. both the `/queue` and the `/backup` directories should have some combined maximum storage threshold, which we'll check on some schedule (perhaps every 6 hours). 
3. if they have maxed out the allocated space and there are images in `/backup`, remove **the oldest** images in `/backup` until we're back below the threshold
4. if they have maxed out the allocated space and there are NO images in `/backup`, that likely means there's a long-lasting internet outage and the `/queue` is using all of the available space, so we need to start culling the images in the `/queue` **at random** until we're back below the threshold.

Basically, have some fixed amount of disk space shared between the un-uploaded images in the `/queue` and the already-uploaded images in the `/backup`, and during normal operation we'd be using pretty much all of that space for backing up the most recent images, but if the base goes offline and the queue starts to build up and we need to make more space, prioritize the deletion of the oldest files in `/backup` until we've exhausted all backed up images, then delete from the images in the `/queue` at random.

**Comment by ejcunningham** [comment]
Some thoughts on this issue from an outsider:

Disk space is certainly a classic embedded/remote systems time bomb that is known to brick devices. As a result, my intuition is to prefer the simplest possible solution where fewer things can go wrong, especially if the scenarios in which the simple solution breaks down (e.g. very long connectivity outages) are rare. Congruously, I also find that it's easiest for users to accommodate scenarios that are predictable. 

In this case, a simple solution is to delete images after some (potentially configurable) amount of time, e.g. 30 or 90 days. This can be done with a 1-liner cron job using linux `find`, but is simple enough to implement in the application as well if you don't want to involve external utilities. This could potentially be paired with a heartbeat that can alert users when a machine is offline and/or accumulating a deep queue, which is probably a useful feature regardless. The disk can be conservatively sized to accommodate the desired lookback window (or vice versa), and if the alarm fires, the user knows exactly how long they have to remediate the issue before they start losing data. I wonder if this would be sufficient?

The proposals to gracefully degrade using selective deletion effectively extend the supportable window at the expense of fidelity in a way that's sort of hard to reason about. Surely there are scenarios where this tradeoff makes sense, but it's not clear to me if we are there yet.

**Comment by erictncabramson** [comment]
Backlog review notes:

- Good spike for Falk
- From user standpoint:
    - What is the objective of keeping the images?

**Comment by postfalk** [comment]
@nathanielrindlaub Looking at the stack, I got a question. Is there a separation beween Multibase and animl-base? Or in plain English is there a good way to test animl-base without having without actually having base station and cameras on hand?

What I picture would some setup where I either copy images per hand or create script that creates fake images in a directory that is monitored by animl-base (and does not depend on Buckeye Multibase?

**Comment by nathanielrindlaub** [comment]
@postfalk it's been a long time since I've done local dev/testing on this, but I think you could comment out the [multibase initialization lines](https://github.com/tnc-ca-geo/animl-base/blob/0fe23a3e9bdcd4a2f3f9f7d7424ed59034d08a77/src/app.js#L62C1-L64C23) and it should run on your local machine.

I might be forgetting something, but I think all animl-base does is start multibase applciation and [tail the multibase logs](https://github.com/tnc-ca-geo/animl-base/blob/0fe23a3e9bdcd4a2f3f9f7d7424ed59034d08a77/src/app.js#L66C1-L68C30), parsing certain log messages and reformatting them and sending them to CloudWatch Metrics.

**Comment by postfalk** [comment]
Just a clarification. We are talking about storage (disk) here or do we have memory issues as well?

**Comment by erictncabramson** [comment]
1. How do we purge backlog list when we've reached some threshold
2. How do we handle managing disk space during internet outage in which they're backing up in the queue but haven't been moved to the archive

**Comment by postfalk** [comment]
@nathanielrindlaub Can we assume Node18.js or newer? Or should we be more conservative?

**Comment by nathanielrindlaub** [comment]
@postfalk the setup documentation has the node download link pinned to 18, so I think we can assume 18 or newer.

**Comment by nathanielrindlaub** [comment]
Closing out! Falk's PR can be found [here](https://github.com/tnc-ca-geo/animl-base/pull/52).

All base stations now have disk management and watchdog cron jobs enabled.
